src/vcpkg/base/messages.cpp

```diff
@@ -0,0 +1,283 @@
+            std::vector<StringLiteral> localization_comments; // const after startup
+
+            bool initialized = false;
+            std::vector<std::string> localized_strings;
+        };
+
+        // to avoid static initialization order issues,
+        // everything that needs the messages needs to get it from this function
+        Messages& messages()
+        {
+            static Messages m;
+            return m;
+        }
+    }
+
+    void threadunsafe_initialize_context()
+    {
+        Messages& m = messages();
+        if (m.initialized)
+        {
+            write_unlocalized_text_to_stdout(Color::Error,
+                                 "double-initialized message context; this is a very serious bug in vcpkg\n");
+            Checks::exit_fail(VCPKG_LINE_INFO);
+        }
+        m.localized_strings.resize(m.names.size());
+        m.initialized = true;
+    }
+    void threadunsafe_initialize_context(const Json::Object& message_map)
+    {
+        threadunsafe_initialize_context();
+
+        Messages& m = messages();
+        std::vector<std::string> names_without_localization;
+
+        for (::size_t index = 0; index < m.names.size(); ++index)
+        {
+            const auto& name = m.names[index];
+            if (auto p = message_map.get(m.names[index]))
+            {
+                m.localized_strings[index] = p->string().to_string();
+            }
+            else if (Debug::g_debugging)
+            {
+                // we only want to print these in debug
+                names_without_localization.push_back(name);
+            }
+        }
+
+        if (!names_without_localization.empty())
+        {
+            println(Color::Warning, msgNoLocalizationForMessages);
+            for (const auto& name : names_without_localization)
+            {
+                write_unlocalized_text_to_stdout(Color::Warning, fmt::format("    - {}\n", name));
+            }
+        }
+    }
+
+    void threadunsafe_initialize_context(const Filesystem& fs, StringView language, const Path& locale_base)
+    {
+        auto path_to_locale = locale_base;
+        path_to_locale /= language;
+        path_to_locale += ".json";
+
+        auto message_map = Json::parse_file(VCPKG_LINE_INFO, fs, path_to_locale);
+        if (!message_map.first.is_object())
+        {
+            write_unlocalized_text_to_stdout(
+                Color::Error,
+                fmt::format("Invalid locale file '{}' - locale file must be an object.\n", path_to_locale));
+            Checks::exit_fail(VCPKG_LINE_INFO);
+        }
+        threadunsafe_initialize_context(message_map.first.object());
+    }
+
+    ::size_t detail::number_of_messages() { return messages().names.size(); }
+
+    ::size_t detail::startup_register_message(StringLiteral name, StringLiteral format_string, StringLiteral comment)
+    {
+        Messages& m = messages();
+
+        auto name_is_used = std::find(m.names.begin(), m.names.end(), name);
```

Comment:
This is `O(N^2)` -- we should not be doing `O(N^2)` things on every startup. Instead, let's only check this when generating the localization file (which will be part of CI).